src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+
+newtype LoweringState
+  = LoweringState { _nextRef :: AnfRef
+                  -- ^ The next 'AnfRef' to track an existing or new variable
+                  -- usage.
+                  }
+  deriving (Eq, Show)
+
+makeLenses ''LoweringState
+
+-- | Monad transformer stack for lowering from 'Term' to 'Anf'.
+--
+-- Instead of using Reader to thread 'BindingStack' state around, we could
+-- pass around explicit stacks through @convert*@ and continuations; but by
+-- using explicit /transformations/ (with `local`) we eliminate bugs pertaining
+-- to the use of an incorrect stack variable. Additionally this current design
+-- is more amenable to a move to the use of @Cont@, if we decide to go that
+-- route.
+type LoweringM = ReaderT BindingStack (State LoweringState)
+
+-- | Dispenses an 'AnfRef' to track the usage of a 'Term' or newly-introduced
+-- ("anf") variable to let-name a subexpression.
+allocRef :: LoweringM AnfRef
+allocRef = do
+  curr <- use nextRef
+  nextRef.refId %= succ
+  return curr
+
+-- | In converting a 'Term' to ANF, specifies whether the allocated value or
+-- result of (prim/fun/thunk) application should be bound to an 'AnfLet'
+-- corresponding to an existing ('Let') 'TermBinding' or a new 'AnfBinding'.
+data Binding
+  = TermBinding
+  | AnfBinding [AnfRef] -- A list because we use unboxed tuples and 2D binders.
+  deriving (Eq, Show)
+
+-- | Bumps the provided source variable by the number of intermediate lets that
+-- have been introduced since the source binder this variable points to.
+translateTermVar :: Variable -> BindingStack -> Variable
+translateTermVar var@(GlobalVarSym _) _ = var
+translateTermVar var@(LocalVar lnv) stack =
+  case mIndirections of
+    Nothing ->
+      adjust displacement var
+    Just indirections ->
+      adjust (displacement + depth) $ indirections V.! slot
+
+  where
+    depth = lnv ^. lnDepth
+    slot  = lnv ^. lnSlot.slotIndex.to fromIntegral
+
+    levels :: [BindingLevel]
+    levels  = take (fromIntegral $ depth + 1) stack
+
+    -- Calculate the displacement sum (from the number of let-introductions) and
+    -- grab the last (possible) indirection in a single pass:
+    (displacement, mIndirections) = (getSum *** join . getLast) . mconcat $
+      (Sum . _levelIntros &&& Last . Just . view levelIndirections) <$> levels
+
+    adjust :: Word32 -> Variable -> Variable
+    adjust offset = localNameless.lnDepth +~ offset
+
+-- | Updates the top of the 'BindingStack' to account for the existing
+-- (pre-translated) 'Term' 'Variable's.
+trackVariables :: Binding
+               -> [Variable]
+               -- ^ Existing source variables that've already been translated to
+               -- point an appropriate number of binders up in ANF land.
+               -> BindingStack
+               -> BindingStack
+trackVariables (AnfBinding refs) vars stack =
+  stack & _head.levelRefs %~ addRefs
+  where
+    addRefs m = foldl' (flip $ uncurry Map.insert) m $ zip refs vars
```

Comment:
I wonder if `addRefs = Map.union (Map.fromList (zip refs vars))` is better? I think it's a little easier to read. The efficiency is at least as good.